sound/soc/sof/core.c

```diff
@@ -353,24 +343,30 @@ int snd_sof_device_remove(struct device *dev)
 	if (IS_ENABLED(CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE))
 		cancel_work_sync(&sdev->probe_work);
 
-	snd_sof_fw_unload(sdev);
-	snd_sof_ipc_free(sdev);
-	snd_sof_free_debug(sdev);
-	snd_sof_free_trace(sdev);
+	if (!(sdev->probe_error_mask & SOF_FW_PROBE_ERROR_FW_UNLOAD))
+		snd_sof_fw_unload(sdev);
+	if (!(sdev->probe_error_mask & SOF_FW_PROBE_ERROR_IPC_FREE))
+		snd_sof_ipc_free(sdev);
+	if (!(sdev->probe_error_mask & SOF_FW_PROBE_ERROR_FREE_DEBUG))
+		snd_sof_free_debug(sdev);
+	if (!(sdev->probe_error_mask & SOF_FW_PROBE_ERROR_FREE_TRACE))
+		snd_sof_free_trace(sdev);
 
 	/*
 	 * Unregister machine driver. This will unbind the snd_card which
 	 * will remove the component driver and unload the topology
 	 * before freeing the snd_card.
 	 */
 	snd_sof_machine_unregister(sdev, pdata);
+
 	/*
 	 * Unregistering the machine driver results in unloading the topology.
 	 * Some widgets, ex: scheduler, attempt to power down the core they are
 	 * scheduled on, when they are unloaded. Therefore, the DSP must be
 	 * removed only after the topology has been unloaded.
 	 */
-	snd_sof_remove(sdev);
+	if (!(sdev->probe_error_mask & SOF_FW_PROBE_ERROR_DSP_REMOVE))
+		snd_sof_remove(sdev);
```

Comment:
unrelated to this patch, but just wondering: is the comment above actually true? I don't see `hda_dsp_remove()` doing topology unloading, and that's a standard implementation of `.remove()`?